File: go-backend-interview-prep/07_io_files_networking/07_csv_basics/main.go
package main

import (
	"encoding/csv"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

func main() {
	// --- Example 1: read CSV from file ---
	fmt.Println("=== Read sample.csv ===")
	f, err := os.Open("07_io_files_networking/07_csv_basics/sample.csv")
	if err != nil {
		f, err = os.Open("sample.csv")
	}
	if err != nil {
		fmt.Println("  could not open sample.csv:", err)
		return
	}
	defer f.Close()

	reader := csv.NewReader(f)
	records, err := reader.ReadAll()
	if err != nil {
		fmt.Println("  read error:", err)
		return
	}

	header := records[0]
	fmt.Println("  header:", header)
	for _, row := range records[1:] {
		fmt.Printf("  row: name=%s age=%s city=%s\n", row[0], row[1], row[2])
	}

	// --- Example 2: read from string ---
	fmt.Println("\n=== Read CSV from string ===")
	csvData := "product,price\n\"Widget, Large\",9.99\nGadget,4.50"
	r := csv.NewReader(strings.NewReader(csvData))
	for {
		row, err := r.Read()
		if err != nil {
			break
		}
		fmt.Printf("  %v\n", row)
	}

	// --- Example 3: write CSV ---
	fmt.Println("\n=== Write CSV ===")
	dir, _ := os.MkdirTemp("", "csv")
	defer os.RemoveAll(dir)

	outPath := filepath.Join(dir, "output.csv")
	outFile, _ := os.Create(outPath)

	writer := csv.NewWriter(outFile)
	writer.Write([]string{"name", "score"})
	writer.Write([]string{"alice", "95"})
	writer.Write([]string{"bob", "87"})
	writer.Write([]string{"charlie, jr.", "92"}) // comma in field -- auto-quoted
	writer.Flush()
	outFile.Close()

	if err := writer.Error(); err != nil {
		fmt.Println("  write error:", err)
	}

	data, _ := os.ReadFile(outPath)
	fmt.Printf("  written CSV:\n%s", data)
}

File: go-backend-interview-prep/07_io_files_networking/07_csv_basics/main.py
"""CSV basics -- Python equivalent of the Go example."""

import csv
import io
import os
import shutil
import tempfile


def main() -> None:
    # --- Example 1: read CSV from file ---
    print("=== Read sample.csv ===")
    try:
        f = open("07_io_files_networking/07_csv_basics/sample.csv")
    except FileNotFoundError:
        f = open("sample.csv")

    with f:
        reader = csv.reader(f)
        header = next(reader)
        print(f"  header: {header}")
        for row in reader:
            if row:  # skip empty rows
                print(f"  row: name={row[0]} age={row[1]} city={row[2]}")

    # --- Example 1b: DictReader (Python bonus) ---
    print("\n=== DictReader (Python-specific) ===")
    try:
        f = open("07_io_files_networking/07_csv_basics/sample.csv")
    except FileNotFoundError:
        f = open("sample.csv")

    with f:
        for row in csv.DictReader(f):
            print(f"  {row['name']} is {row['age']} in {row['city']}")

    # --- Example 2: read from string ---
    print("\n=== Read CSV from string ===")
    csv_data = 'product,price\n"Widget, Large",9.99\nGadget,4.50'
    reader = csv.reader(io.StringIO(csv_data))
    for row in reader:
        print(f"  {row}")

    # --- Example 3: write CSV ---
    print("\n=== Write CSV ===")
    tmpdir = tempfile.mkdtemp()
    out_path = os.path.join(tmpdir, "output.csv")

    try:
        with open(out_path, "w", newline="") as f:
            writer = csv.writer(f)
            writer.writerow(["name", "score"])
            writer.writerow(["alice", "95"])
            writer.writerow(["bob", "87"])
            writer.writerow(["charlie, jr.", "92"])  # comma -- auto-quoted

        with open(out_path) as f:
            print(f"  written CSV:\n{f.read()}", end="")
    finally:
        shutil.rmtree(tmpdir, ignore_errors=True)


if __name__ == "__main__":
    main()
